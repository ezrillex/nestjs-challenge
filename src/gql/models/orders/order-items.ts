import { Field, ID, Int, ObjectType } from '@nestjs/graphql';
import { Orders } from './orders';
import { ProductVariations } from '../products/product-variations/product-variations';

@ObjectType()
export class OrderItems {
  @Field(() => ID)
  id: string;

  @Field(() => Orders)
  order: Orders;

  @Field(() => ProductVariations)
  product_variation: ProductVariations;

  @Field(() => Int)
  quantity: number;

  @Field(() => Number)
  price_purchased_at: number;
}
