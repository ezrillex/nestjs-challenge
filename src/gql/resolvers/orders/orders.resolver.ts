import { Args, Context, Mutation, Query, Resolver } from '@nestjs/graphql';
import { ProductsService } from '../../../services/products/products.service';
import { RequiresRole } from '../../../decorators/requires-role/requires-role.decorator';
import { roles } from '@prisma/client';
import { ParseUUIDPipe } from '@nestjs/common';

@Resolver()
export class OrdersResolver {
  // todo refactor this into its separate service?
  constructor(private productsService: ProductsService) {}

  @RequiresRole(roles.customer)
  @Mutation(() => String, { nullable: true })
  async createOrder(@Context('req') request: Request) {
    // let's assume the customer is buying the entire cart.
    return this.productsService.CreateOrder(request['user'].id);
  }

  // @RequiresRole(roles.customer)
  // @Mutation(() => String, { nullable: true })
  // async removeLikeProduct(
  //   @Args('like_id', { type: () => String }, ParseUUIDPipe)
  //   like_id: string,
  // ) {
  //   return this.productsService.RemoveLike(like_id);
  // }
  //
  // @RequiresRole(roles.customer)
  // @Query(() => [LikesOfProducts], { nullable: true })
  // async getLikes(@Context('req') request: Request) {
  //   return this.productsService.GetLikes(request['user'].id);
  // }
}
