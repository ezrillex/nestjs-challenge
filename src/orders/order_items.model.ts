import { Field, ID, Int, ObjectType } from '@nestjs/graphql';
import { ProductVariations } from '../products/product_variation/product-variations.model';

@ObjectType()
export class OrderItems {
  @Field(() => ID)
  id: string;

  // circular reference issue. omitting orders field.
  // @Field(() => Orders, { nullable: true })
  // order: Orders;

  @Field(() => ProductVariations)
  product_variation: ProductVariations;

  @Field(() => Int)
  quantity: number;

  @Field(() => Number)
  price_purchased_at: number;
}
