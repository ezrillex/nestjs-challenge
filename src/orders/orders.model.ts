import { OrderStatus, PaymentStatus } from '@prisma/client';
import { OrderItems } from './order_items.model';
import { Field, ID, ObjectType } from '@nestjs/graphql';
import { Users } from '../auth/users.model';

@ObjectType()
export class Orders {
  @Field(() => ID)
  id: string;

  @Field(() => Users)
  user: Users;

  // todo when implementing resolve field maybe add this
  //payments: [PaymentIntents!]
  //webhooks: [IncomingPaymentWebhooks!]

  @Field(() => String)
  paymentStatus: PaymentStatus;

  @Field(() => String)
  orderStatus: OrderStatus;

  //orderStatusHistory: [OrderAudit!]!

  @Field(() => [OrderItems])
  order_items: OrderItems[];
}
