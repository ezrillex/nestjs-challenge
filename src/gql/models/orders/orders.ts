import { OrderStatus, PaymentStatus } from '@prisma/client';
import { OrderItems } from './order-items';
import { Field, ID, ObjectType } from '@nestjs/graphql';
import { Users } from '../users/users/users';

@ObjectType()
export class Orders {
  @Field(() => ID)
  id: string;

  @Field(() => Users)
  user: Users;

  //payments: [PaymentIntents!]
  //webhooks: [IncomingPaymentWebhooks!]
  //orderDetails: [OrderDetails!]!

  @Field(() => PaymentStatus)
  paymentStatus: PaymentStatus;

  @Field(() => OrderStatus)
  orderStatus: OrderStatus;

  //orderStatusHistory: [OrderAudit!]!

  @Field(() => [OrderItems])
  OrderItems: OrderItems[];
}
