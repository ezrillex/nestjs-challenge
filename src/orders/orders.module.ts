import { Module } from '@nestjs/common';
import { OrdersService } from './orders.service';
import { OrdersResolver } from './orders.resolver';
import { PrismaService } from '../prisma/prisma.service';
import { UsersService } from '../users/users.service';
import { OrderItemsResolver } from './order-items/order-items.resolver';
import { ProductsService } from '../products/products.service';
import { StripeService } from '../payments/stripe.service';

@Module({
  providers: [
    OrdersService,
    OrdersResolver,
    PrismaService,
    UsersService,
    OrderItemsResolver,
    ProductsService,
    StripeService,
  ],
})
export class OrdersModule {}
