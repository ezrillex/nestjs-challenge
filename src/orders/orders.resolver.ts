import {
  Args,
  Context,
  Mutation,
  Parent,
  Query,
  ResolveField,
  Resolver,
} from '@nestjs/graphql';
import { RequiresRole } from '../common/decorators/requires-role.decorator';
import { roles } from '@prisma/client';
import { ParseUUIDPipe } from '@nestjs/common';
import { Orders } from './orders.model';
import { OrdersService } from './orders.service';
import { UsersService } from '../users/users.service';
import { StripeService } from '../payments/stripe.service';
import { GetOrdersInput } from './inputs/get_orders.input';
import { Users } from '../users/users.model';
import { OrderItems } from './order-items/order_items.model';
import { PaymentIntents } from '../payments/payments.model';

@Resolver(() => Orders)
export class OrdersResolver {
  constructor(
    private readonly ordersService: OrdersService,
    private readonly usersService: UsersService,
    private readonly stripeService: StripeService,
  ) {}

  @RequiresRole(roles.customer)
  @Mutation(() => Orders, { nullable: true })
  async createOrder(@Context('req') request: Request): Promise<Orders> {
    // let's assume the customer is buying the entire cart.
    return this.ordersService.createOrder(request['user'].id);
  }

  @Query(() => [Orders], { nullable: true })
  async getOrders(
    @Args('GetOrdersInput') getOrdersInput: GetOrdersInput,
    @Context('req') request: Request,
  ): Promise<Orders[]> {
    return this.ordersService.getOrders(
      request['user'].id,
      request['user'].role,
      getOrdersInput,
    );
  }

  @Query(() => Orders, { nullable: true })
  async getOrder(
    @Args('order_id', { type: () => String }, ParseUUIDPipe)
    order_id: string,
    @Context('req') request: Request,
  ): Promise<Orders> {
    return this.ordersService.getOrder(
      order_id,
      request['user'].id,
      request['user'].role,
    );
  }

  @ResolveField()
  async user(@Parent() orders: Orders): Promise<Users> {
    const { id } = orders;
    return this.usersService.getUserByOrder(id);
  }

  @ResolveField()
  async order_items(@Parent() cart_items: Orders): Promise<OrderItems[]> {
    const { id } = cart_items;
    return this.ordersService.getOrderItemsByOrder(id);
  }

  @ResolveField()
  async payments(@Parent() cart_items: Orders): Promise<PaymentIntents[]> {
    const { id } = cart_items;
    return this.stripeService.getPaymentsByOrder(id);
  }
}
