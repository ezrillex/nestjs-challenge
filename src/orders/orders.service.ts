import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
} from '@nestjs/common';
import { roles } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { Orders } from './orders.model';
import { GetOrdersInput } from './inputs/get_orders.input';
import { OrderItems } from './order-items/order_items.model';

@Injectable()
export class OrdersService {
  constructor(private readonly prisma: PrismaService) {}

  async CreateOrder(user_id: string): Promise<Orders> {
    const cart_items = await this.prisma.cartItems.findMany({
      select: {
        id: true,
        quantity: true,
        product_variation_id: true,
        product_variation: {
          select: {
            price: true,
          },
        },
      },
      where: {
        user_id: user_id,
        product_variation: {
          product: { is_deleted: false, is_published: true },
        },
      },
    });
    if (cart_items.length === 0) {
      throw new BadRequestException('User has no items in cart!');
    }

    const order_items = cart_items.map((item) => ({
      quantity: item.quantity,
      product_variation_id: item.product_variation_id,
      price_purchased_at: item.product_variation.price,
    }));
    const cart_ids = cart_items.map((item) => item.id);

    const create_query = {
      data: {
        user: {
          connect: { id: user_id },
        },
        order_items: {
          createMany: {
            data: order_items,
          },
        },
      },
    };

    const delete_query = {
      where: {
        id: {
          in: cart_ids,
        },
      },
    };

    try {
      return await this.prisma.$transaction(async () => {
        const create_result = await this.prisma.orders.create(create_query);
        const delete_result =
          await this.prisma.cartItems.deleteMany(delete_query);
        if (delete_result.count === 0) {
          throw new InternalServerErrorException('Error deleting cart items!');
        }
        return create_result;
      });
    } catch {
      throw new InternalServerErrorException(
        'Transaction failed: An error occurred when creating the order!',
      );
    }
  }

  async GetOrders(
    user_id: string,
    role: roles,
    inputs: GetOrdersInput,
  ): Promise<Orders[]> {
    const find_parameters = { where: {} };
    const pagination = { skip: inputs.offset ?? 0, take: inputs.first ?? 10 };

    switch (role) {
      case 'customer':
        find_parameters.where = {
          user_id: user_id,
          // don't hide deleted or private items.
        };
        break;
      case 'manager':
        if (inputs.client_id) {
          find_parameters.where = {
            user_id: inputs.client_id,
            // don't hide deleted or private items.
          };
        }
        break;
      default:
        throw new InternalServerErrorException(
          'Invalid role expected customer or manager',
        );
    }

    return this.prisma.orders.findMany({ ...find_parameters, ...pagination });
  }

  async GetOrder(
    order_id: string,
    client_id: string,
    role: roles,
  ): Promise<Orders> {
    const filter = {
      id: order_id,
    };
    // if not a manager is only allowed to view its own orders.
    if (role !== roles.manager) {
      filter['user_id'] = client_id;
    }
    const find_parameters = {
      where: { ...filter },
    };

    // todo should I error if there is nothing ? if not from self user it returns null is it graphql way of doing it like so?
    return this.prisma.orders.findUnique(find_parameters);
  }

  async ResolveOrderItemsField(order_id: string): Promise<OrderItems[]> {
    const { order_items } = await this.prisma.orders.findUnique({
      where: {
        id: order_id,
      },
      select: {
        order_items: true,
      },
    });
    return order_items;
  }
}
