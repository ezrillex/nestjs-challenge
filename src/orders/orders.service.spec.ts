import { Test, TestingModule } from '@nestjs/testing';
import { OrdersService } from './orders.service';
import { PrismaService } from '../prisma/prisma.service';
import {
  CartItems,
  roles,
  ProductVariations,
  Orders,
  Prisma,
} from '@prisma/client';

describe('Orders Service', () => {
  let service: OrdersService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PrismaService, OrdersService],
    }).compile();

    service = module.get<OrdersService>(OrdersService);
    prismaService = module.get<PrismaService>(PrismaService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should have method CreateOrder', () => {
    expect(service.CreateOrder).toBeDefined();
  });

  it('should have method GetOrders', () => {
    expect(service.GetOrders).toBeDefined();
  });

  it('should have method GetOrder', () => {
    expect(service.GetOrder).toBeDefined();
  });

  describe('Get Orders', () => {
    it('should query the user id if role customer', async () => {
      let data;
      jest
        .spyOn(prismaService.orders, 'findMany')
        .mockImplementation((query) => {
          data = query;
          return null;
        });

      await expect(
        service.GetOrders('the_user_id', roles.customer),
      ).resolves.not.toThrow();

      expect(data).toMatchSnapshot('role customer result');
    });

    it('should query the client id if role manager', async () => {
      let data;
      jest
        .spyOn(prismaService.orders, 'findMany')
        .mockImplementation((query) => {
          data = query;
          return null;
        });

      await service.GetOrders('the_user_id', roles.manager, 'client_id');

      expect(data).toMatchSnapshot('role manager result');
    });

    it('should throw if the client id is null and  if role manager', async () => {
      jest.spyOn(prismaService.orders, 'findMany').mockImplementation(() => {
        return null;
      });

      await expect(
        service.GetOrders('the_user_id', roles.manager),
      ).rejects.toThrowErrorMatchingSnapshot('client id is null');
    });
  });

  describe('Get One Order', () => {
    it('should include order id and client in query', async () => {
      let data;
      jest
        .spyOn(prismaService.orders, 'findUnique')
        .mockImplementation((query) => {
          data = query;
          return null;
        });

      await service.GetOrder('testing_user_id', 'testing_client_id');

      expect(data).toMatchSnapshot('get one order result');
    });
  });

  describe('Create an order', () => {
    it('should throw if no cart items', async () => {
      jest.spyOn(prismaService.cartItems, 'findMany').mockResolvedValue([]);

      await expect(
        service.CreateOrder('testing_user_id'),
      ).rejects.toThrowErrorMatchingSnapshot('error no items in the cart');
    });

    it('should unroll nested product info and send a valid query', async () => {
      type QueryType = CartItems & {
        product_variation: ProductVariations;
      };

      jest.spyOn(prismaService.cartItems, 'findMany').mockResolvedValue([
        {
          quantity: 1,
          product_variation_id: 'prod var 1',
          product_variation: {
            id: 'some id',
            title: 'test title',
            price: 34,
            stock: 5,
            product_id: 'some product id',
            last_updated_by: 'some user id',
            last_updated_at: new Date(),
          },
        },
        {
          quantity: 4,
          product_variation_id: 'prod var 2',
          product_variation: {
            id: 'some id 2',
            title: 'test title2',
            price: 343,
            stock: 53,
            product_id: 'some product id2',
            last_updated_by: 'some user id2',
            last_updated_at: new Date(),
          },
        },
      ] as unknown as QueryType[]);

      let result;
      jest.spyOn(prismaService.orders, 'create').mockImplementation((input) => {
        result = input;
        return null;
      });
      await expect(
        service.CreateOrder('testing_user_id'),
      ).rejects.toThrowErrorMatchingSnapshot(
        'error ocurred when creating order',
      );
      expect(result).toMatchSnapshot('query of order creation test');
    });

    it('should unroll cart ids and send a valid query', async () => {
      type QueryType = CartItems & {
        product_variation: ProductVariations;
      };

      jest.spyOn(prismaService.cartItems, 'findMany').mockResolvedValue([
        {
          id: 'some id',
          quantity: 1,
          product_variation_id: 'prod var 1',
          product_variation: {
            title: 'test title',
            price: 34,
            stock: 5,
            product_id: 'some product id',
            last_updated_by: 'some user id',
            last_updated_at: new Date(),
          },
        },
        {
          id: 'some id 2',
          quantity: 4,
          product_variation_id: 'prod var 2',
          product_variation: {
            title: 'test title2',
            price: 343,
            stock: 53,
            product_id: 'some product id2',
            last_updated_by: 'some user id2',
            last_updated_at: new Date(),
          },
        },
      ] as unknown as QueryType[]);

      jest.spyOn(prismaService.orders, 'create').mockResolvedValue({
        id: '1',
      } as unknown as Prisma.Prisma__OrdersClient<Orders>);

      const spy = jest
        .spyOn(prismaService.cartItems, 'deleteMany')
        .mockResolvedValue({
          count: 1,
        });

      const returned = await service.CreateOrder('testing_user_id');

      const input = spy.mock.calls[0];

      expect(input).toMatchSnapshot('query of delete many test');
      expect(returned).toMatchSnapshot('create order return');
    });
    it('throws error if delete count is 0, error', async () => {
      type QueryType = CartItems & {
        product_variation: ProductVariations;
      };

      jest.spyOn(prismaService.cartItems, 'findMany').mockResolvedValue([
        {
          id: 'some id',
          quantity: 1,
          product_variation_id: 'prod var 1',
          product_variation: {
            title: 'test title',
            price: 34,
            stock: 5,
            product_id: 'some product id',
            last_updated_by: 'some user id',
            last_updated_at: new Date(),
          },
        },
        {
          id: 'some id 2',
          quantity: 4,
          product_variation_id: 'prod var 2',
          product_variation: {
            title: 'test title2',
            price: 343,
            stock: 53,
            product_id: 'some product id2',
            last_updated_by: 'some user id2',
            last_updated_at: new Date(),
          },
        },
      ] as unknown as QueryType[]);

      jest.spyOn(prismaService.orders, 'create').mockResolvedValue({
        id: '1',
      } as unknown as Prisma.Prisma__OrdersClient<Orders>);

      jest.spyOn(prismaService.cartItems, 'deleteMany').mockResolvedValue({
        count: 0,
      });

      await expect(
        service.CreateOrder('testing_user_id'),
      ).rejects.toThrowErrorMatchingSnapshot('error unexpected when deleting');
    });
  });
});
