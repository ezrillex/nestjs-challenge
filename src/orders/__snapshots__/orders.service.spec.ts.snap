// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders Service Create an order should throw if no cart items: error no items in the cart 1`] = `"User has no items in cart!"`;

exports[`Orders Service Create an order should unroll cart ids and send a valid query: create order return 1`] = `"1"`;

exports[`Orders Service Create an order should unroll cart ids and send a valid query: query of delete many test 1`] = `
[
  {
    "where": {
      "id": {
        "in": [
          "some id",
          "some id 2",
        ],
      },
    },
  },
]
`;

exports[`Orders Service Create an order should unroll nested product info and send a valid query: error ocurred when creating order 1`] = `"An error occurred when creating the order!"`;

exports[`Orders Service Create an order should unroll nested product info and send a valid query: query of order creation test 1`] = `
{
  "data": {
    "order_items": {
      "createMany": {
        "data": [
          {
            "price_purchased_at": 34,
            "product_variation_id": "prod var 1",
            "quantity": 1,
          },
          {
            "price_purchased_at": 343,
            "product_variation_id": "prod var 2",
            "quantity": 4,
          },
        ],
      },
    },
    "user": {
      "connect": {
        "id": "testing_user_id",
      },
    },
  },
  "include": {
    "order_items": true,
    "user": true,
  },
}
`;

exports[`Orders Service Create an order throws error if delete count is 0, error: error unexpected when deleting 1`] = `"An unexpected error occurred!"`;

exports[`Orders Service Get One Order should include order id and client in query: get one order result 1`] = `
{
  "where": {
    "id": "testing_user_id",
    "user_id": "testing_client_id",
  },
}
`;

exports[`Orders Service Get Orders should query the client id if role manager: role manager result 1`] = `
{
  "where": {
    "user_id": "client_id",
  },
}
`;

exports[`Orders Service Get Orders should query the user id if role customer: role customer result 1`] = `
{
  "where": {
    "user_id": "the_user_id",
  },
}
`;

exports[`Orders Service Get Orders should throw if the client id is null and  if role manager: client id is null 1`] = `"No client id for manager specific request!"`;
